cosmwasm/packages/wasmi-runtime/src/exports.rs

```diff
@@ -116,6 +116,10 @@ pub unsafe extern "C" fn ecall_init(
     sig_info: *const u8,
     sig_info_len: usize,
 ) -> InitResult {
+    if let Err(err) = recursion_depth::increment() {
+        error!("recursion limit exceeded, can not perform query!");
```

Comment:
You could make a contract that inits itself during the init, so I guess so.

Reply:
@Cashmaney initializing a new contract will only happen after the first init completes, and that doesn't count towards recursion.
@assafmo Not sure what you mean by your question, but I believe that this error condition is currently unreachable. I think we can safely remove it completely right now, and have `recursion_depth::increment()` simply increment the counter with no further checks, but i wanted to stay on the safe side here, in case something changes in the future, and we can easily spot that we forgot to add a limit somewhere.

Reply:
we should rephrase the error message as `"can not perform init"` ,as @Cashmaney suggests

Reply:
Yep. I'd add a code comment with what you just wrote:
> I believe that this error condition is currently unreachable. I think we can safely remove it completely right now, and have `recursion_depth::increment()` simply increment the counter with no further checks, but i wanted to stay on the safe side here, in case something changes in the future, and we can easily spot that we forgot to add a limit somewhere.

Reply:
@Cashmaney the goal here is to prevent more than 5 VMs allocated at the same time.

The only 3 possible combinations are:

- init + query + query + query + query
- handle + query + query + query + query
- query + query + query + query + query